File: Scala/euler/src/re/_1r/euler/problems/108.scala
package re._1r.euler.problems

import scala.concurrent._
import ExecutionContext.Implicits.global

object Euler108 {
  def gcd(a: Int, b: Int): Int = {
    if (a == 0) b else if (b == 0) a else {
      gcd(b, a % b)
    }
  }
  def recip(n: Int) = {
    (n + 1 to 2 * n).count(v => {
      val lcm = n * v / gcd(n, v)
      1d / v + 1d / lcm == 1d / n
    })
  }
  def result = {
    def top = 100
    val f = Array(
      Future((1 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((2 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((3 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((4 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((5 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((6 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((7 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((8 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((9 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((10 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((11 to Int.MaxValue by 12).find(recip(_) > top).get),
      Future((12 to Int.MaxValue by 12).find(recip(_) > top).get)
    )
    while(!f.exists(_.isCompleted)) {}
    f.find(_.isCompleted).get.value.get.get
  }
}

File: Elixir/108.exs
defmodule Euler108 do
  def recip(n) do
    Enum.count(n + 1..2 * n, fn v ->
      lcm = n * v / Integer.gcd(n, v)
      1 / v + 1 / lcm == 1 / n
    end)
  end
  def result do
    Enum.find(1..9999999999, fn n ->
      x = recip(n)
      #IO.puts("#{n} -> #{x}")
      x > 100
    end)
  end
end

begin = System.monotonic_time(:nanosecond)
result = Euler108.result
fin = System.monotonic_time(:nanosecond)
IO.puts("#{result}\ntook: #{(fin - begin)/1000000000}")
